### PR TITLE
Migrate to Postgres

### LINKED ISSUE #490: Migrate to Postgres
postgres + drizzle

### PR DESCRIPTION
**Pull request by js0mmer**
## Description

* Migrated to postgres
* added drizzle
* refactored routes
  * made sure we are determining userid from session, not letting user dictate it
* refactored login checks on the frontend with a hook
* added course bag routes
* added separate course bag slice

## todo
- [ ] need to clean up mongoose stuff before merging
- [x] update docs

## Steps to verify/test this change:

- [ ] Verify changes work as expected on staging instance
- [ ] make sure the site still works

## Final Checks:

- [ ] Verify successful deployment

(optional)

- [ ] Write tests
- [ ] Write documentation

## Issues

Closes #490

### CODE REVIEW

File: api/src/controllers/reviews.ts
Comment (Awesome-E) [inline api/src/controllers/reviews.ts:68]: This seems a bit clunky. At the very least we don't have to construct an array just to spread it. If anything, I'd prefer to declare `const filters = []` and simply add each `eq()` to the filters array, then just do `.and(...filters)` at the end.

```suggestion
        ...(courseId ? [eq(review.courseId, courseId)] : []),
        ...(professorId ? [eq(review.professorId, professorId)] : []),
        ...(userId ? [eq(review.userId, userId)] : []),
        ...(reviewId ? [eq(review.id, reviewId)] : []),
        ...(verified !== undefined ? [eq(review.verified, verified)] : []),
```

File: api/src/controllers/reviews.ts
Comment (Awesome-E) [inline api/src/controllers/reviews.ts:130]: Ideally this check happens before they even open the form – let's create an issue for that

File: api/src/controllers/reviews.ts
Comment (Awesome-E) [inline api/src/controllers/reviews.ts:217]: Put as a variable instead of as a comment

`const featuredReviewCriteria = [desc(review.content), desc(voteSubQuery.score), desc(review.verified)`

`.orderBy(...featuredReviewCriteria)`

File: api/src/controllers/reviews.ts
Comment (Awesome-E) [inline api/src/controllers/reviews.ts:224]: It looks like converting dates is repeated behavior. We should have some sort of helper function to automatically convert all date key/value pairs into ISO strings

File: api/src/controllers/roadmap.ts
Comment (Awesome-E) [inline api/src/controllers/roadmap.ts:40]: if `planner` is an import that you don't want to override, I'd rather have this variable name be `plannerData`

File: api/src/controllers/users.ts
Comment (Awesome-E) [inline api/src/controllers/users.ts:20]: same as previous occurrences of this pattern

File: api/src/db/migrate.ts
Comment (Awesome-E) [inline api/src/db/migrate.ts:36]: I know it's nice to be able to directly return the object, but at the same time repeating `sessions.session!.passport!.user!` is a bit clunky

If you _really_ want to, you could also do `.map(({ session: { session: { passport: { user } } }) => ({...})` but that's equally as confusing

Comment (Awesome-E) [reply api/src/db/migrate.ts:36]: I think it'd be best to just do something like this:

```ts
.map(session => {
  const { id, name, email, picture } = session.session!.passport!.user!
  return { googleId: id, name, email, picture }
})
```

File: api/src/db/migrate.ts
Comment (Awesome-E) [inline api/src/db/migrate.ts:75]: similar to dates the other way, we should have a helper to construct this pattern
`dateLookup('lastRoadmapEditAt', roadmap.roadmap.timestamp)`

File: api/src/db/schema.ts
Comment (Awesome-E) [inline api/src/db/schema.ts:118]: is there a reason for using `sql'...'` here instead of something like `or()`?

File: api/src/models/review.ts
Comment (Awesome-E) [inline api/src/models/review.ts:73]: why is this added in this PR?

File: api/src/models/session.ts
Comment (Awesome-E) [inline api/src/models/session.ts:1]: similar to previous file – why are the mongoose schemas still here?

File: site/src/component/SideInfo/SideInfo.tsx
Comment (Awesome-E) [inline site/src/component/SideInfo/SideInfo.tsx:1]: Lint warning on changed file – I'm indifferent about changing it here

File: site/src/helpers/planner.ts
Comment (Awesome-E) [inline site/src/helpers/planner.ts:95]: I don't like this type of pattern, but not sure if there's a better way

File: site/src/pages/RoadmapPage/ImportTranscriptPopup.tsx
Comment (Awesome-E) [inline site/src/pages/RoadmapPage/ImportTranscriptPopup.tsx:10]: should we consider renaming this to `quarterNames` in `@peterportal/types` altogether since it's just names and not JSON?

File: site/src/pages/RoadmapPage/Transfer.tsx
Comment (Awesome-E) [inline site/src/pages/RoadmapPage/Transfer.tsx:58]: I assume this is because `null` was giving an error?

**Review by Awesome-E — CHANGES_REQUESTED** [review]

File: site/src/pages/RoadmapPage/Transfer.tsx
Comment (js0mmer) [reply site/src/pages/RoadmapPage/Transfer.tsx:58]: yes, database returns a real number or null for this but the value attribute can only be number or undefined

File: api/src/db/migrate.ts
Comment (js0mmer) [reply api/src/db/migrate.ts:36]: i don't really care about the code in this file because i'm not merging it, this just for a one-time use of migrating everything over

File: api/src/models/session.ts
Comment (js0mmer) [reply api/src/models/session.ts:1]: i'll delete them before merging, I needed them for the migration script

File: site/src/component/SideInfo/SideInfo.tsx
Comment (js0mmer) [reply site/src/component/SideInfo/SideInfo.tsx:1]: i might make a pr in the future addressing all the remaining warnings

File: api/src/controllers/reviews.ts
Comment (js0mmer) [reply api/src/controllers/reviews.ts:130]: Ah, so you mean like doing the check to prevent them from even having the option on the form to create an identical review?

File: api/src/controllers/reviews.ts
Comment (js0mmer) [reply api/src/controllers/reviews.ts:68]: Yes, I'll remove the outer array. I'm really not sure if adding will look any cleaner though since we would just be adding based on these same conditions so it would look kind of the same. I took the spreading idea from [drizzle docs](https://orm.drizzle.team/docs/select#conditional-select).

File: api/src/controllers/reviews.ts
Comment (js0mmer) [reply api/src/controllers/reviews.ts:224]: On it

File: site/src/pages/RoadmapPage/ImportTranscriptPopup.tsx
Comment (js0mmer) [reply site/src/pages/RoadmapPage/ImportTranscriptPopup.tsx:10]: Yeah, the type actually comes from peterportal-api-next-types though which is exported by `@peterportal/types`

File: api/src/db/schema.ts
Comment (js0mmer) [reply api/src/db/schema.ts:118]: I think `or()` works here? I have to use a non-null assertion for some reason but I think it's working.
```ts
    voteCheck: check('votes_vote_check', or(eq(table.vote, 1), eq(table.vote, -1))!),
```

File: api/src/controllers/reviews.ts
Comment (Awesome-E) [reply api/src/controllers/reviews.ts]: I think there is a way to make the helper type-safe. I'm not used to writing complicated type-defs like this, but I think that's a good use for ChatGPT.

File: api/src/db/schema.ts
Comment (Awesome-E) [reply api/src/db/schema.ts]: Great, I think that's better for consistency

File: site/src/pages/RoadmapPage/ImportTranscriptPopup.tsx
Comment (Awesome-E) [reply site/src/pages/RoadmapPage/ImportTranscriptPopup.tsx:10]: I see. In that case, we might want to rename the export

File: api/src/controllers/reviews.ts
Comment (js0mmer) [reply api/src/controllers/reviews.ts]: Cooked something up with some help from github copilot

File: site/src/pages/RoadmapPage/ImportTranscriptPopup.tsx
Comment (js0mmer) [reply site/src/pages/RoadmapPage/ImportTranscriptPopup.tsx]: Added an export for `quarterNames`

File: api/src/controllers/reviews.ts
Comment (js0mmer) [reply api/src/controllers/reviews.ts]: Ok I found another reference on the [drizzle docs](https://orm.drizzle.team/docs/guides/conditional-filters-in-query) and I was able to remove the spread operators.

**Review by Awesome-E — APPROVED** [review]
Just create the issues for the two unresolved comments, otherwise looks pretyt much good to me! Wouldn't hurt to get a second opinion from @timobraz

File: .github/workflows/build-and-deploy.yml
Comment (js0mmer) [reply .github/workflows/build-and-deploy.yml:1]: #499

File: api/src/controllers/reviews.ts
Comment (js0mmer) [reply api/src/controllers/reviews.ts:130]: #500

File: api/src/db/schema.ts
Comment (js0mmer) [reply api/src/db/schema.ts:118]: for future reference, `or()` does not work here. we have to use `sql` for checks

**Pull request merged**
